### PR TITLE
Retry create issue to cope with duplicate keys

### PR DESCRIPTION
**Pull request by guillep2k**
This is a workaround for the potential race condition in `models/issue.go` (#7887).

Sadly, there seems to be no universal way of telling a duplicate key from other kinds of errors, so the way I did it was to assume all errors can be retried.

I could write some small util with hardcoded error number detection as suggested [here](http://go-database-sql.org/errors.html). For instance, we could do this for MySQL:

```
if driverErr, ok := err.(*mysql.MySQLError); ok { // Now the error number is accessible directly
	if driverErr.Number == 1045 {
		// Handle the permission-denied error
	}
}
```

It doesn't look so good. Some examples:

#### MySQL
Error number: `1169`
Message: `Can't write, because of unique constraint, to table '%s'`

#### MSSQL
Error number: `2601`
Message: `Cannot insert duplicate key row in object '%.*ls' with unique index '%.*ls'.`
_also_
Error number: `2627`
Message: `Violation of %ls constraint '%.*ls'. Cannot insert duplicate key in object '%.*ls'.`

#### PostgreSQL
Error number: `23505`
Message: `duplicate key value violates unique constraint "%s"`

You get the idea.

### CODE REVIEW

**Comment by zeripath** [comment]
I'm not sure this will completely solve the issues if e is a session in a transaction for two reasons:

* I suspect a failed insert in the transaction will cause rollback of the entire transaction. 
* You might not actually get the error until the transaction tries to commit because it's at that point that the constraints are fully tested. 

I think you have to put the retry at the router level.

**Comment by lunny** [comment]
Maybe we could add test for this to confirm it.

**Comment by guillep2k** [comment]
> 
> 
> Maybe we could add test for this to confirm it.

That would be a problem. I couldn't figure out a good way of unit testing this, other than adding some ugly hook in newIssue() to force the error. I've been looking around but didn't find anything convincing. Suggestions are welcomed.

PS: This is the "ugly hooks" solution I thought of (**definitely not pretty**):

```
// There's no good way to identify a duplicate key error in database/sql; brute force some retries
dupIndexAttempts := issueMaxDupIndexAttempts
for {
	maxIndex, err := getMaxIndexOfIssue(e, opts.Issue.RepoID)
	if err != nil {
		return err
	}
	opts.Issue.Index = maxIndex + 1
	// Milestone and assignee validation should happen before insert actual object.
	// *********************************************************************
	_, err = e.Insert(opts.Issue)
        err = testHookShouldFail(err) // **** MY EYES, MY EYES! ****
        if err == nil {
		break
	}
	// *********************************************************************
	dupIndexAttempts--
	if dupIndexAttempts <= 0 {
		return err
	}
}
```

**Comment by guillep2k** [comment]
> * I suspect a failed insert in the transaction will cause rollback of the entire transaction.
> * You might not actually get the error until the transaction tries to commit because it's at that point that the constraints are fully tested.

@zeripath I think this is not the case: for many verbs (e.g. `DELETE`, `UPDATE`) you get the number of affected rows immediately after execution, `INSERT` does that as well, but its usage is less obvious (e.g. `INSERT` with conditions, or `INSERT` from `SELECT`). By looking into xorm code, the Insert is ultimately translated to a `session.exec(sqlStr, args...)`, so I don't think there's any buffering involved.

**Comment by typeless** [comment]
I thought we should not pass the issue id explicitly, but rather letting the database do the assignment for us. Is that possible (and sane)?

**Comment by guillep2k** [comment]
> 
> 
> I thought we should not pass the issue id explicitly, but rather letting the database do the unique id generation for us. Is that possible (and sane)?

@typeless This is not the ID value (which the database does set with an auto-increment) but the correlative issue number for that repository. i.e issue 1, issue 2, issue 3, push request 4, issue 5, etc.

**Comment by typeless** [comment]
@guillep2k oh, I see. So it's not the primary key but still required to be unique right?

**Comment by guillep2k** [comment]
@typeless Yes, and they should be consecutive too (i.e. not random). Since it's calculated by running `max(index)`, there's a statistical chance that two instances could attempt the same "next" number and collide. That's the reason for the retry loop, which is not very elegant but might scale better than other alternatives.

**Comment by typeless** [comment]
I tried to add an integration test for it.
```
func TestNewIssueConcurrent(t *testing.T) {

        prepareTestEnv(t)
        session := loginUser(t, "user2")

        f := func(ctx context.Context, cancel context.CancelFunc, wg *sync.WaitGroup) {
                defer wg.Done()

                for i := 0; i < 200; i++ {
                        select {
                        case <-ctx.Done():
                                return
                        default:
                                issueUrl := testNewIssue(t, session, "user2", "repo1", fmt.Sprintf("Title %d", i), "Description")
                                if issueUrl == "" {
                                        cancel()
                                        return
                                }
                        }
                }
        }

        wg := &sync.WaitGroup{}
        ctx, cancel := context.WithCancel(context.Background())
        defer cancel()

        wg.Add(2)
        go f(ctx, cancel, wg)
        go f(ctx, cancel, wg)
        wg.Wait()
}
```
it turns out the `newIssue` operation is quite slow, so probably not suitable to be included in the test suite. And sometimes it would hang unexpectedly. It might be a bug in the test or another undiscovered bug.

**Comment by guillep2k** [comment]
> 
> 
> I tried to add an integration test for it.
> 

@typeless Very interesting. The problem I see is that this might or might not trigger the issue even before the PR, and that will depend a lot on server load, speed, luck, database used, etc. It won't be very reliable.

On mssql I could use some trigger to produce the error using the session's [CONTEXT_INFO](https://docs.microsoft.com/en-us/sql/t-sql/functions/context-info-transact-sql?view=sql-server-2017) and a `BEFORE INSERT` trigger, so we _could_ test the functionality in the integrations tests as a special case. This would have the advantage that the "ugliness" will reside in the test and not in the actual code. However, I think connection pooling might be a problem and I can't reliably test on all mssql versions.

PostgreSQL has the [`current_setting()`](https://stackoverflow.com/questions/17956601/check-for-value-with-current-setting) function that also looks promising.

**Comment by sapk** [comment]
Can we make the id calculation during the insert by using a subrequest and retrieving the assigned id later ? Doing so the lock for the insert will be used for the subrequest for the id calculation and prevent any reading during the whole process.

**Comment by typeless** [comment]
[this](https://stackoverflow.com/a/34572959) SO answer gives some insight.
We should be able to calculate the serial number per repo from the other fields: (id, Repo_id).

**Comment by guillep2k** [comment]
> 
> Can we make the id calculation during the insert by using a subrequest and retrieving the assigned id later ? Doing so the lock for the insert will be used for the subrequest for the id calculation and prevent any reading during the whole process.

@sapk  You mean in the tests or in the main code? I don't know if that's possible without resorting to engine specific features.

**Comment by guillep2k** [comment]
Now that I think of it, in plain SQL, I'd do:
```
insert into issue (repoid, index, title, body) values (4, (select max(index)+1 from issue as i2 where i2.repoid = 4), 'Some title', 'Some text');
```
This should be atomic in the database unless _chaos isolation_ is selected, even if no transaction is in place. Downside is the direct SQL statement. I'll make another PR testing this.

**Comment by sapk** [comment]
@guillep2k that the kind of SQL query I was thinking and even if it a raw sql query it is a simple solution. 
The SQL trigger suggested by @typeless could also be a good solution but I never tested across all SQL engine we support (it seems supported at least) and it should be setup at the db init.

**Comment by sapk** [comment]
For testing, you can re-use/inpire you from the pattern test that @typeless implemented for a sqlite lib improvement https://github.com/mattn/go-sqlite3/pull/439/files#diff-bd96d50843020c6a7e1423b5b913d554R121 and call multiple time `newIssue`

**Comment by guillep2k** [comment]
> 
> 
> @guillep2k that the kind of SQL query I was thinking and even if it a raw sql query it is a simple solution.
> The SQL trigger suggested by @typeless could also be a good solution but I never tested across all SQL engine we support (it seems supported at least) and it should be setup at the db init.

I'm messing around with xorm to make this work. I'll probably have to ask @lunny for some changes there so I can add an override expression to the inserted value (xorm currently doesn't let me do that, but has some boilerplate for it). Something like:
```
func (session *Session) InsertWithSubExpr(bean interface{}, exprlist map[string]*Builder) ...
```
Then I could do:
```
subexpr := make(map[string]*builder.Builder)
subexpr["Index"] = builder.Select("MAX(`index`) + 1").From("`issue`").Where(builder.Eq{"`repo_id`": opts.Issue.RepoID})
_, err = e.InsertWithSubExpr(opts.Issue, subexpr)
```
or something around those lines.

**Comment by guillep2k** [comment]
OK, so I've made several tests and it looks like the most robust way to do this would be with a manual insert. Something like:

```
INSERT INTO `issue` (`repo_id`, `poster_id`, ... ,`name`, `content`, ... ,`index`)
        SELECT $1, $2, ..., $14, $15, ..., MAX(`index`) + 1 FROM `issue` WHERE `repo_id` = $1;
```
Where `$1`, `$2`, etc. are the positional parameters. Then the value of `index` must be retrieved by querying the database back with `getIssueByID()` or something similar.

This is not directly supported by xorm, but I think all databases will handle this method correctly (here are some fiddlers for [MySQL](http://sqlfiddle.com/#!9/c007cf/2), [MSSQL](http://sqlfiddle.com/#!18/c007c/1), [Oracle](http://sqlfiddle.com/#!4/c007c/1), [PostgreSQL](http://sqlfiddle.com/#!17/c007c/1), [SQLite](http://sqlfiddle.com/#!5/c007c/1)).

This should save us from retrying an insert on the edge case of a collision, but at the cost of performing a double query (insert + select) in all cases. It is however a neater solution than others I've played with.

Problems:

* It's not easy to get hard information about the atomicity of an `INSERT AS SELECT` statement for all databases; it _should_ be atomic in all cases, however.
* Without a patch on xorm to handle this directly, the SQL statement should be constructed in the Gitea code. Xorm [has some helpers](https://github.com/go-xorm/builder#insert) to avoid engine specific syntax to get in the way, so the problem is reduced to properly maintaining the column to field association when building the SQL statement.
* One thing that xorm does not help with (not without a patch) is the auto-increment columns, i.e. `Issue.ID`. Each engine has its own quirks for handling them (`LAST_INSERT_ID()`, `RETURNING ID`, parsing the text output...). Since the `INSERT` would be handled by an `Exec()`, retrieving the inserted `ID` should be taken care of manually. Having this code on Gitea would be very clumsy.

So, in my opinion this problem has a relatively good solution, but it should be handled by the ORM and not by Gitea. Other than that, the "retry hack" proposed by @lunny looks like a reasonable compromise.

**Comment by guillep2k** [comment]
Opened an issue in xorm for this, in case it is considered a valuable addition:
https://gitea.com/xorm/builder/issues/58

**Comment by lunny** [comment]
@sapk @guillep2k https://github.com/go-xorm/xorm/pull/1401

**Comment by guillep2k** [comment]
> 
> 
> @sapk @guillep2k @typeless [go-xorm/xorm#1401](https://github.com/go-xorm/xorm/pull/1401)

That's great, @lunny. I've tested the new feature with this and it worked perfectly. This is the code generated for SQLITE:

```
2019/08/22 12:21:16 .../xorm/session_raw.go:178:exec() [I] [SQL] INSERT INTO `issue` (`repo_id`, `poster_id`, `original_author`, `original_author_id`, `name`, `content`, `milestone_id`, `priority`, `is_closed`, `is_pull`, `num_comments`, `ref`, `deadline_unix`, `created_unix`, `updated_unix`, `closed_unix`, `is_locked`, `index`) SELECT ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, coalesce(MAX(`index`),0)+1 FROM `issue` WHERE (repo_id=?) []interface {}{1, 1, "", 0, "This is a pull title", "", 0, 0, false, true, 0, "", 0, 1566487276, 1566487276, 0, false, 1} - took: 1.240578ms
```

**Comment by guillep2k** [comment]
Superseded by #7944

**Comment by guillep2k** [comment]
Reopen due to #7944 failing.

File: models/issue.go [hunk 1]
```diff
@@ -1108,9 +1103,24 @@ func newIssue(e *xorm.Session, doer *User, opts NewIssueOptions) (err error) {
 		}
 	}
 
-	// Milestone and assignee validation should happen before insert actual object.
-	if _, err = e.Insert(opts.Issue); err != nil {
-		return err
+	// There's no good way to identify a duplicate key error in database/sql; brute force some retries
+	dupIndexAttempts := issueMaxDupIndexAttempts
+	for {
```
Comment (sapk) [inline models/issue.go]: If we go back this way. We should define a max number of tries and trigger en error if we doesn't find a good number.

**Comment by sapk** [comment]
Can you add the tests in this PR to validate ?

**Comment by guillep2k** [comment]
> 
> 
> Can you add the tests in this PR to validate ?

@sapk, I don't think I should, since this PR does not eliminate the problem but only mitigates/reduces it. Chances are that integrations will start to fail for everybody in a random fashion because of the added test.

**Comment by guillep2k** [comment]
### Follow up for the interested
I don't know if there's anybody left here interested in alternatives to solve the potential duplicate key problem, but if this PR (retry insert in case of error) is not convincing, I've found another way to solve this problem that's testable, scalable, universal and reliable... but conceptually more complicated. 

It might be an over-engineered solution. I'm just being thorough by considering all the edges. This PR (retrying insert in case of error) might be all that we need, so... I leave it to you guys to decide if we need another solution.

### The other solution
The short story is:

First, `issue.index` must be `NULLABLE`.

We insert the issue with `NULL` in the `index` column. This will never cause a duplicate key error because `NULL` values do not get indexed:
```
INSERT INTO issue (repo_id, index, ...) values ({repoID}, NULL, ...) RETURNING id
```
The engine will return the value for `id` as usual.

After the insert, we calculate the proper value for index with the following query:
```
SELECT COALLESCE(MAX(index),0) + COUNT(*) - COUNT(index) AS "index_value"
FROM issue WHERE repo_id = {repoID} and id <= {ID}
```
Then we update the issue with the calculated value:
```
UPDATE issue SET index = {index_value} WHERE id = {ID}
```
If anybody is interested I can explain this concept with more detail, but I must warn you: it's a long read.

Advantages:

* Universal: all databases will support it.
* Does not generate locks or depend on them.
* Can be used without transactions.
* The _concept_ is easily unit testable **without stressing the server** (but must be tested independently of the `newIssue()` function).
* It will never generate duplicates.

Disadvantages:

* It's hard to follow.
* It requires two updates for each issue (`INSERT` + `UPDATE`).
* If something interrupts the process in the middle (e.g. power failure), between the `INSERT` and the `UPDATE`, the issue will be left with a `NULL` index (but the process can be resumed at Gitea's startup with no problem).
* Queries that scan the `issues` table should be aware of this process by filtering out rows that have a `NULL` in the `index` column to avoid getting mutating results.

I can work on this concept if it raises enough interest, but I'm afraid I'm just boring you by now.

**Comment by guillep2k** [comment]
I've combined the `x.SetExpr().Where().Insert()` method with the multiple retries, and removed unnecessary code related to the index calculation. This should greatly reduce the incidence of errors because the `SELECT` is now closer to the `INSERT` in the database. But the retries are still necessary because of locks problems as shown in #7959.

**Review by sapk — APPROVED** [review]
Your suggestion could be review in an other PR. But in the meantime, this pr feel a good enough solution to move forward.

**Comment by sapk** [comment]
Make LGTM work.

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
We still need a test for creating issues on multiple go routines.

**Comment by lunny** [comment]
The file name should be `issue_stress_test.go`, otherwise it will not be run I think.

**Comment by sapk** [comment]
I know it is know the perfect solution but what about putting them as a benchmark ? It will have a loop for multiple execution. It will not block CI but can easily be tested against new solution.

**Comment by sapk** [comment]
No I a mean adding it as a benchmark (they are not run on CI) that are basicly loop of tests. For exemple like the func BenchmarkRepoMigrate.

**Comment by typeless** [comment]
Another way is to use a build TAG such as `stress` to distinguish it from the other tests.
